### ISSUE TITLE
After restarting your teletask central unit, jeletask is not reconnecting properly

### ISSUE THREAD

**Issue by SanderM2** [ridiekel/jeletask#19]
Bug:
If I restart my teletask central unit (for example when I change things in Prosoft) jeletask isn't reconnecting properly.
jeletask has to be restarted in order to reconnect.

Expected behaviour:
jeletask should reconnect to the central unit

**Comment by ridiekel** [comment]
This should be fixed in 4.1.0, can you please confirm?

**Comment by SanderM2** [comment]
Hi @ridiekel
I wonder... did you just add a fix to the binary release, or did you assume this was fixed in the last PR you just merged? Because.. I have not looked in to this myself. If you didn't fix it it's not fixed in 4.1.0 I guess...
I'm asking because I can't find the related change in github's history.

**Comment by ridiekel** [comment]
I thought I saw this fix in your PR, but in hindsight, that was a fix for when te mqtt connection is lost. Not for when TT connection was lost, so I gues this is indeed not fixed yet

**Comment by SanderM2** [comment]
When restarting the central unit the error messages shown are:

First:
```
2022-09-30 08:48:36.611 -TRACE - i.g.r.j.c.b.m.messages.MessageSupport    : Sending message: {"acknowledged":false,"clientConfig":{"host":"192.168.178.210","port":55957,"type":"MICROS_PLUS"}}
2022-09-30 08:48:36.612 -DEBUG - i.g.r.j.client.TeletaskClientImpl        : 
Command: KEEP_ALIVE
Payload: 
	None
Length: 4
Checksum calculation steps: 2 + 4 + 11 + 1 = 18
Raw Bytes: 02 04 0B 01 12
--------------------------------------------
| STX | Length | Command | Param 0 | ChkSm |
--------------------------------------------
| 02  | 04     | 0B      | 01      | 12    |
--------------------------------------------
2022-09-30 08:48:36.612 -TRACE - i.g.r.j.c.b.m.messages.MessageSupport    : Message sent: {"acknowledged":false,"clientConfig":{"host":"192.168.178.210","port":55957,"type":"MICROS_PLUS"}}
2022-09-30 08:48:38.653 -ERROR - i.g.r.j.c.b.m.executor.MessageExecutor   : Problem running message: {"acknowledged":false,"clientConfig":{"host":"192.168.178.210","port":55957,"type":"MICROS_PLUS"}}

io.github.ridiekel.jeletask.client.builder.message.messages.AcknowledgeException: KEEPALIVE - Did not receive acknowledge from the Teletask Central Unit within 2000 ms
	at io.github.ridiekel.jeletask.client.builder.message.messages.MessageSupport.waitForAcknowledge(MessageSupport.java:84) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.builder.message.messages.MessageSupport.execute(MessageSupport.java:58) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.builder.message.executor.MessageExecutor.run(MessageExecutor.java:22) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.builder.composer.microsplus.MicrosPlusMessageHandler$MicrosPlusKeepAliveStrategy.execute(MicrosPlusMessageHandler.java:130) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.TeletaskClientImpl$KeepAliveService.lambda$run$0(TeletaskClientImpl.java:443) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
Caused by: org.awaitility.core.ConditionTimeoutException: Condition with alias 'Acknowlegde - KEEPALIVE' didn't complete within 2 seconds because condition with io.github.ridiekel.jeletask.client.builder.message.messages.MessageSupport was not fulfilled.
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:148) ~[awaitility-4.0.1.jar!/:na]
	at org.awaitility.core.CallableCondition.await(CallableCondition.java:78) ~[awaitility-4.0.1.jar!/:na]
	at org.awaitility.core.CallableCondition.await(CallableCondition.java:26) ~[awaitility-4.0.1.jar!/:na]
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:873) ~[awaitility-4.0.1.jar!/:na]
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:842) ~[awaitility-4.0.1.jar!/:na]
	at io.github.ridiekel.jeletask.client.builder.message.messages.MessageSupport.waitForAcknowledge(MessageSupport.java:75) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	... 7 common frames omitted
```


Second:
```
2022-09-30 08:49:36.621 -TRACE - i.g.r.j.c.b.m.messages.MessageSupport    : Sending message: {"acknowledged":false,"clientConfig":{"host":"192.168.178.210","port":55957,"type":"MICROS_PLUS"}}
2022-09-30 08:49:36.622 -DEBUG - i.g.r.j.client.TeletaskClientImpl        : 
Command: KEEP_ALIVE
Payload: 
	None
Length: 4
Checksum calculation steps: 2 + 4 + 11 + 1 = 18
Raw Bytes: 02 04 0B 01 12
--------------------------------------------
| STX | Length | Command | Param 0 | ChkSm |
--------------------------------------------
| 02  | 04     | 0B      | 01      | 12    |
--------------------------------------------
2022-09-30 08:49:36.623 -ERROR - i.g.r.j.c.b.m.executor.MessageExecutor   : Problem running message: {"acknowledged":false,"clientConfig":{"host":"192.168.178.210","port":55957,"type":"MICROS_PLUS"}}

io.github.ridiekel.jeletask.client.TeletaskClientImpl$CommunicationException: Problem sending message to teletask central unit
	at io.github.ridiekel.jeletask.client.TeletaskClientImpl.send(TeletaskClientImpl.java:241) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.builder.message.messages.MessageSupport.execute(MessageSupport.java:54) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.builder.message.executor.MessageExecutor.run(MessageExecutor.java:22) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.builder.composer.microsplus.MicrosPlusMessageHandler$MicrosPlusKeepAliveStrategy.execute(MicrosPlusMessageHandler.java:130) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.TeletaskClientImpl$KeepAliveService.lambda$run$0(TeletaskClientImpl.java:443) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
Caused by: java.net.SocketException: Connection reset by peer
	at java.base/sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:417) ~[na:na]
	at java.base/sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:437) ~[na:na]
	at java.base/sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:823) ~[na:na]
	at java.base/java.net.Socket$SocketOutputStream.write(Socket.java:1035) ~[na:na]
	at java.base/java.io.OutputStream.write(OutputStream.java:127) ~[na:na]
	at io.github.ridiekel.jeletask.client.builder.message.MessageUtilities.send(MessageUtilities.java:101) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.TeletaskClientImpl.send(TeletaskClientImpl.java:239) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	... 7 common frames omitted
```

Next, after these 2 message, it keeps repeating this error until restarted:
```
2022-09-30 08:50:36.618 -TRACE - i.g.r.j.c.b.m.messages.MessageSupport    : Sending message: {"acknowledged":false,"clientConfig":{"host":"192.168.178.210","port":55957,"type":"MICROS_PLUS"}}
2022-09-30 08:50:36.618 -DEBUG - i.g.r.j.client.TeletaskClientImpl        : 
Command: KEEP_ALIVE
Payload: 
	None
Length: 4
Checksum calculation steps: 2 + 4 + 11 + 1 = 18
Raw Bytes: 02 04 0B 01 12
--------------------------------------------
| STX | Length | Command | Param 0 | ChkSm |
--------------------------------------------
| 02  | 04     | 0B      | 01      | 12    |
--------------------------------------------
2022-09-30 08:50:36.619 -ERROR - i.g.r.j.c.b.m.executor.MessageExecutor   : Problem running message: {"acknowledged":false,"clientConfig":{"host":"192.168.178.210","port":55957,"type":"MICROS_PLUS"}}

io.github.ridiekel.jeletask.client.TeletaskClientImpl$CommunicationException: Problem sending message to teletask central unit
	at io.github.ridiekel.jeletask.client.TeletaskClientImpl.send(TeletaskClientImpl.java:241) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.builder.message.messages.MessageSupport.execute(MessageSupport.java:54) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.builder.message.executor.MessageExecutor.run(MessageExecutor.java:22) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.builder.composer.microsplus.MicrosPlusMessageHandler$MicrosPlusKeepAliveStrategy.execute(MicrosPlusMessageHandler.java:130) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.TeletaskClientImpl$KeepAliveService.lambda$run$0(TeletaskClientImpl.java:443) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
Caused by: java.net.SocketException: Broken pipe
	at java.base/sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:417) ~[na:na]
	at java.base/sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:437) ~[na:na]
	at java.base/sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:823) ~[na:na]
	at java.base/java.net.Socket$SocketOutputStream.write(Socket.java:1035) ~[na:na]
	at java.base/java.io.OutputStream.write(OutputStream.java:127) ~[na:na]
	at io.github.ridiekel.jeletask.client.builder.message.MessageUtilities.send(MessageUtilities.java:101) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	at io.github.ridiekel.jeletask.client.TeletaskClientImpl.send(TeletaskClientImpl.java:239) ~[jeletask-client-4.0.4.jar!/:4.0.4]
	... 7 common frames omitted
```

**Comment by ridiekel** [comment]
It probably is something with exception handling not handling this exeption (and pbb others also) correctly. I haven't looked at it since until recently I was the only one using this project :-)

Since now it has more impact I will put it in higher prio to actually test this. I hope to have some time this weekend. This thogether with the great work you already did will make this project more usable and stable.

Thank for the help, and keep up the good work!

**Comment by SanderM2** [comment]
I just took a quick look (on my tablet while on train...)

It looks like restart() is only called from the KeepAliveService run() method. It should log a message like "Exception ({}) caught in KeepAliveService: {}" but that message never showed up in my logs.

Instead, we see "Problem sending message to teletask central unit" which happens if send() on the socket fails.

So I think that the exception in KeepAliveService run() method never happens because the exception was already taken care of in send() ?

**Comment by SanderM2** [comment]
I added this to my jeletask version: https://github.com/ridiekel/jeletask/commit/3bb21d34428f27b9d9e689d3825eb6b827d434c9

It now successfully reconnects to the teletask central after rebooting it.

**Comment by ridiekel** [comment]
Just released your merge request in 4.1.1, can you test if it works with the released version?

**Comment by SanderM2** [comment]
Yes, it works.
This issue can be closed :)
